### DISCUSSION TITLE
[docs] NestJS serialization vs MikroORM serialization

### DISCUSSION THREAD

**Discussion by marcpicaud** [mikro-orm/nestjs#65] [Ideas]
# tldr; 

> As a MikroORM newbie using NestJS, I want to find in the docs a clear warning about the class-transformer incompatibility, so that I don't get stuck trying to follow NestJS serialization techniques.

------

First of all, thanks for this amazing project.

I think we could improve the documentation regarding one aspect : **serialization**.

[NestJS documentation about serialization](https://docs.nestjs.com/techniques/serialization) is pushing developers towards [class-transformer](https://github.com/typestack/class-transformer) and interceptors.

In the context of MikroORM, this is misleading. Concepts like `Collection`, `Reference` etc are not compatible with class-transformer, and people who are following NestJS serialization techniques by the book are getting stuck (well, I was).

I think that this incompatibility is not an issue at all, the built-in serialization in MikroORM is equivalent. But the NestJS integration should display a clear warning about that.

Related links : 
- https://github.com/mikro-orm/mikro-orm/issues/720
- https://github.com/mikro-orm/mikro-orm/discussions/2687 
- https://github.com/mikro-orm/nestjs/pull/38

What do you guys think ? Happy to submit a PR.

**Comment by B4nan** [comment]
PR welcome 👍

Note that serialization with class-transformer is possible, you just need to return the DTO, not entity (so `wrap(e).toJSON()`). Creating new entities from DTO is not possible, that needs to happen via `em.create()` or similar, but again, you can use class-transformer before you call `em.create()` to do any kind of validation on the DTO.

Maybe we could have some kind of extension for class-transformer?

**Reply by marcpicaud** [reply]
> Note that serialization with class-transformer is possible, you just need to return the DTO, not entity (so wrap(e).toJSON()).

`wrap(e).toJSON()` returns a generic object, so  [Nest's ClassSerializerInterceptor](https://github.com/nestjs/nest/blob/master/packages/common/serializer/class-serializer.interceptor.ts#L77) will do nothing with it before sending back the HTTP response. This interceptor expects a class instance (with decorated properties like `@Exclude()`, `@Expose()` etc.).

The problem is not really with class-transformer, but more with ClassSerializerInterceptor (which is widely used).

**Comment by marcpicaud** [comment]
Ok I'll make a PR 👍 

**DTO ---> Entity :**

For me there's no issue here in the NestJS context. It provides a `ValidationPipe` so that incoming DTOs are actually validated **before** they reach the API route handlers. Once there, creating the entity through `em.create()` and the incoming DTO is pretty straightforward.

**Entity ---> DTO :**

Here's my understanding of the possible solutions : 
- ~~As you suggested, developers can return the DTO instead of the wrapped entity. Easy to forget though.~~ EDIT: I could not make this work
- We can also drop the `ClassSerializerInterceptor` and just use MikroORM serialization options in the `@Property` decorator. I prefer this way, there are less moving pieces and it does the job just fine (hide, expose, and transform properties)
- Find a way to automatically call `.toJSON()` on wrapped entities  before the `ClassSerializerInterceptor` steps in. Maybe another NestJS interceptor upstream?

**Reply by B4nan** [reply]
> Find a way to automatically call .toJSON() on wrapped entities before the ClassSerializerInterceptor steps in. Maybe another NestJS interceptor upstream?

That's probably what https://github.com/mikro-orm/nestjs/pull/38 is for.
